test/runtime.spec.ts

```diff
@@ -212,3 +212,63 @@ describe('map', () => {
     return true;
   });
 });
+
+describe('filterDeep', () => {
+  jsc.property('returns empty array when no matches', jsc.json, async dict => {
+    const filtered = await filterDeep(dict, (_n: any): _n is string => false);
+    expect(filtered).toEqual([]);
+    return true;
+  });
+
+  jsc.property('returns full object in array when already matches root', jsc.json, async dict => {
+    const filtered = await filterDeep(dict, (_n: any): _n is string => true);
+    expect(filtered).toEqual(dict);
+    return true;
+  });
+
+  jsc.property(
+    'matches array filter with one dimensional arrays',
+    jsc.array(jsc.oneof([jsc.asciistring, jsc.integer, jsc.bool])),
+    async arr => {
+      const filtered = await filterDeep(arr, (n: any): n is string => _.isString(n));
+      expect(filtered).toEqual(arr.filter(item => _.isString(item)));
+      return true;
+    }
+  );
+
+  jsc.property(
+    'matches Object.values.filter with one dimensional objects',
+    jsc.dict(jsc.oneof([jsc.asciistring, jsc.integer, jsc.bool])),
+    async dict => {
+      const filtered = await filterDeep(dict, (n: any): n is string => _.isString(n));
+      expect(filtered).toEqual(Object.values(dict).filter(item => _.isString(item)));
+      return true;
+    }
+  );
+
+  jsc.property(
+    'matches single primitives',
+    jsc.oneof([jsc.asciistring, jsc.integer, jsc.bool]),
+    async item => {
+      const filtered = await filterDeep(item, (n: any): n is string => _.isString(n));
+      expect(filtered).toEqual(_.isString(item) ? [item] : []);
+      return true;
+    }
+  );
+
+  it('returns all nested values with the property', () => {
```

Comment:
would be nice to have similar tests with nested arbitrary data structures with jsverify's `jsc.json`, but I can't really think of how to write the property for that without practically reimplementing the function. :D The "true for all" and "true for none" properties with `jsc.json` I have above already.

Reply:
One way to test this would be to use the equality 
```
const found = []
runtime.map(value, predicate, a=> {  found.push(a); return a })
expect(runtime.filterDeep(value, predicate)).toEqual(found)
```

where value would be any generated json and predicate could be isString, isObject etc

Reply:
I think the above would suffice :/, generally one way to test this kind of things is the generate two value
 - a json object haystack
 - a value Array<string|number> which you use to _.set a known object to the haystack

and then you just assert that you find the known object

Reply:
sounds good!